File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head or k == 1:
            return head

        dummy = ListNode(0, head)
        prev = dummy
        curr = head
        tail = dummy

        length = 0
        while curr:
            length += 1
            curr = curr.next

        while length >= k:
            curr = prev.next
            tail = prev
            for _ in range(k):
                tail = tail.next
            next_group_start = tail.next
            tail.next = None

            prev.next = self.reverse_list(curr)
            curr.next = next_group_start
            prev = curr

            length -= k

        return dummy.next

    def reverse_list(self, head: ListNode) -> ListNode:
        prev = None
        curr = head

        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        return prev

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    if (!head || k === 1) {
        return head;
    }

    let dummy = new ListNode(0, head);
    let prev = dummy;
    let tail = dummy;
    let curr = head;

    let length = 0;
    while (curr) {
        length++;
        curr = curr.next;
    }

    while (length >= k) {
        curr = prev.next;
        tail = prev;
        for (let i = 0; i < k; ++i) {
            tail = tail.next;
        }
        let nextGroupStart = tail.next;
        tail.next = null;

        prev.next = reverseList(curr);
        curr.next = nextGroupStart;
        prev = curr;

        length -= k;
    }

    return dummy.next;
};

function reverseList(head) {
    let prev = null;
    let curr = head;

    while (curr) {
        let nextNode = curr.next;
        curr.next = prev;
        prev = curr;
        curr = nextNode;
    }

    return prev;
}
